[Draft] Blog: Rethinking Software Variants (#1362)
betterscientificsoftware/bssw.io: Articles/Blog/SoftwareVariants.md

Testing gets more complicated: full recompilations are needed of vast parts of the software stack if one cannot create a single environment that enables and tests all functionality, because it might conflict.
Even with sufficient automation, the latter increases continuous integration time and deployment resources.

Typical problems shall be illustrated in a few examples.

### Breaking Public API

A common issues from quickly scattered `#ifdef` switches are changing API interfaces: extra parameters in functions, different class signatures and constructors, varying members.
In the worst case, the change of the upstream option has to be mirrored 1:1 downstream with `#ifdef`s at call locations, e.g., because existing signatures are *changed*.

Such API changes also always lead to incompatible application binary interfaces (ABIs).
Likewise, it is easy to introduce breaking ABI changes that are more subtle, which will result in linkage failures for a new variant at best and subtle breakage at runtime in the worst case.
A typical example is the manipulation of the number of member variables of a public class based on a binary variant.

> 

Author: I can add a section here on actual ABI breaks, e.g., adding a private member to a public class at the wrong location

Author: done

### HPC examples

The most common pattern of variations for HPC software is options for parallelism.
In the simplest case, this is a flag that enables or disables MPI support in the software.
With the rise of GPU computing over the past decade, a similar option is often added for GPU acceleration.

Library and application developers in HPC might then add more domain-specific compilation options, such as additional numerical solvers supported by external math libraries (e.g., BLAS, LAPACK, or FFT) and choices of geometry of a model.

### The catch

Generally, such configuration options can be implemented at build time in two ways: binary switches and multivariant options.
Binary switches are simply turning a functionality on or off, for example, MPI support.
Multivariant options might be more involved, for example, accelerating code for a specific GPU programming environment (CUDA, HIP, SYCL, OpenMP, or OpenACC offloading).

The catch with such options lies in sharing software with other people who want to reuse it in their own libraries and applications, develop against it, and deploy their own product to other developers and users.

> Our experience is that problems for developers of ``downstream" HPC software arise mainly from the implementation of these options.

Author: Is that the correct formatting in `.md` for double quotes? Looks LaTeX-y :)
```suggestion
Our experience is that problems for developers of "downstream" HPC software arise mainly from the implementation of these options.
```

The third-party software can be built with MPI enabled and now introduces a compile-time dependency on MPI signatures even though the specific downstream translation unit never uses it.

The results include breaking builds, the need to communicate additional, potentially inaccurate (unused) dependencies, and breakage in most desktop package managers.
A typical example is Debian and HDF5: one cannot install an MPI-parallel HDF5 package for development and the popular, serial HDFView package at the same time.

**The unconditional MPI initialize (or expectation thereof)**

This is a variation of the previous problem, which occurs at runtime as a result of an MPI binary variant.
If the MPI-enabled variant of the software *expects* that an MPI context will always be provided (or can be established), this breaks serial -- and non-MPI parallelized -- software applications.

One can make the same case for any other runtime that needs to be initialized or finalized, such as initialization of GPU devices or GPU streams.

**On-node acceleration**

> Going into more detail on the previous point, a cardinal pattern in HPC software is to define mutually exclusive binary patterns for the ``acceleration backend" of software.

Author: Is that the correct formatting in `.md` for double quotes? Looks LaTeX-y :)
```suggestion
Going into more detail on the previous point, a cardinal pattern in HPC software is to define mutually exclusive binary patterns for the "acceleration backend" of software.
```

Many single-source performance-portability implementations currently compile to exactly one on-node acceleration backend at a time.

For example, one might compile a numerical package to run with a CUDA backend or compile it again with an OpenMP backend or with a HIP/ROCm backend.

Delegating this decision to compile time takes away the choice to deploy binary packages or pre-build unified containers that could run on various GPU vendor hardware or be run on a CPU or GPU as a runtime option, depending on user need.
It also hinders dependent developments from utilizing the CPU and GPU at the same time, for example, in simulations with dynamic load balancing.

So-called fat binary artifacts can address this problem in part, by compiling multiple backends into the same executable and delegating the code path to choose at runtime.
Unfortunately, few established conventions and tooling currently exist for such an approach, especially across different vendors, so this particular case can be hard to deal with.

### Possible solutions and development policies

> The solution to these challenges starts with everyone thinking of themselves as ``upstream" developers and thinking about how downstream users might reuse their software.

Author: Is that the correct formatting in `.md` for double quotes? Looks LaTeX-y :)
```suggestion
The solution to these challenges starts with everyone thinking of themselves as "upstream" developers and thinking about how downstream users might reuse their software.
```

> This even includes application developers: someone might come up with a clever way to integrate an application -- like a library -- into a larger context, for exsample, for optimizating ensemble use cases or AI/ML workflows.

Author: ```suggestion
This even includes application developers: someone might come up with a clever way to integrate an application -- like a library -- into a larger context, for example, for optimizing ensemble use cases or AI/ML workflows.
```

### Package managers and deployment

But hey, don't we have package managers to solve this problem for us by keeping track of the right variant?
Well, in part.
Of course, modern package managers such as [Spack](https://spack.io) enable developers to define various development environments with exactly chosen and combined variants of software.
This approach is helpful for correct results.
Yet, if one needs to use multiple (compile-time) features of a software, this results in a combinatorial explosion of artifacts for development environments and deployments.

On top of that, many popular package managers in use by developers and users are not as powerful as Spack.
Most do not support binary variants at all -- and are adding them as afterthoughts via package naming extensions, again resulting in combinatorial installs and requiring complex conflict resolutions.

Indeed, independent of whether a package manager supports binary variants well, it is tremendously helpful if a package manager with great dependency resolution can just switch *ON* all options that are *potentially* useful for a system at the same time.

> With thst capability, there are fewer modules to build, no environment switching is needed for development, and binary caches can be smaller.

Author: ```suggestion
With that capability, there are fewer modules to build, no environment switching is needed for development, and binary caches can be smaller.
```

<img src='../../images/Blog_2209_SoftwareVariants_Spack.png' />[Some of the WarpX compile-time options exposed in the Spack package manager.]

### Hands-on examples

**C/C++**

The following design patterns can be used for C/C++ code.

**Header:** Selected opt-in variants should add *extra* header files that expose additional classes and API calls and expose functionality such as MPI-enabled signatures.

> This means at least one public ``facade" header is needed per configuration option, to avoid ``polluting" base functionality with third-party includes.

Author: Is that the correct formatting in `.md` for double quotes? Looks LaTeX-y :)
```suggestion
This means at least one public "facade" header is needed per configuration option, to avoid "polluting" base functionality with third-party includes.
```

        "On-node, accelerated computing backend")

    foreach(backend IN LISTS App_Backend)
        # ...
    endforeach()
    ```

**Python and Fortran**

The following design patterns can be used for Python and Fortran code.

**Modules:** Providing extra (sub)modules that expose functionality that expects or creates a certain runtime context, e.g., MPI-parallelism

**Properties:** Adding properties to the base module to allow querying which opt-in functions are available at runtime

**Modifications of ECP WarpX**

Over the past two years, we redesigned most binary options of the Exascale Computing Project application [WarpX](https://ecp-warpx.github.io) based on these experiences and insights.
We changed binary options that control additional field-solvers that require FFTs and linear algebra to provide additional runtime options when enabled, and without changing application behavior if compiled and not used.

More complicated are changes in simulation geometry.
Ideally, WarpX developers would like to offer users a single deployment that provides 1D, 2D, 3D and quasi-cylindrical (RZ) geometry at the same time.
In AMReX, this is a compile-time option.

> We are addressing this by progressively compiling all geometries ``as 3D" and adding additional 1D and 2D calls to parallel kernel primitives in AMReX 3D interfaces.

Author: Is that the correct formatting in `.md` for double quotes? Looks LaTeX-y :)
